gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/InstrumentatedgoogleHadoopFSInputStream.java

```
@@ -0,0 +1,132 @@
+import com.google.cloud.hadoop.gcsio.GoogleCloudStorageReadOptions;
+import java.io.IOException;
+import java.net.URI;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.statistics.IOStatistics;
+import org.apache.hadoop.fs.statistics.IOStatisticsSource;
+
+/** A class to update the input stream statistics of {@link GoogleHadoopFSInputStreamBase} */
+public class InstrumentatedgoogleHadoopFSInputStream extends GoogleHadoopFSInputStreamBase
+    implements IOStatisticsSource {
+
+  // Statistic tracker of the Input stream
+  private final GhfsInputStreamStatistics streamStatistics;
+
+  // IO Statistics of the current input stream
+  private final IOStatistics ioStatistics;
+
+  InstrumentatedgoogleHadoopFSInputStream(
+      GoogleHadoopFileSystemBase ghfs,
+      URI gcsPath,
+      GoogleCloudStorageReadOptions readOptions,
+      FileSystem.Statistics statistics)
+      throws IOException {
+    super(ghfs, gcsPath, readOptions, statistics);
+    this.streamStatistics =
+        ((InstrumentatedGoogleHadoopFileSystem) ghfs)
+            .getInstrumentation()
+            .newInputStreamStatistics(statistics);
+    this.ioStatistics = streamStatistics.getIOStatistics();
+  }
+
+  InstrumentatedgoogleHadoopFSInputStream(
+      GoogleHadoopFileSystemBase ghfs, FileInfo fileInfo, FileSystem.Statistics statistics)
+      throws IOException {
+    super(ghfs, fileInfo, statistics);
+    this.streamStatistics =
+        ((InstrumentatedGoogleHadoopFileSystem) ghfs)
+            .getInstrumentation()
+            .newInputStreamStatistics(statistics);
+    this.ioStatistics = streamStatistics.getIOStatistics();
+  }
+
+  @Override
+  public synchronized int read() throws IOException {
+    streamStatistics.readOperationStarted(getPos(), 1);
+    int response = 0;
+    try {
+      response = super.read();
+    } catch (IOException e) {
+      streamStatistics.readException();
+      streamStatistics.readOperationCompleted(1, response);
```

we are capturing the readOperationCompleted twice here, and consuming the IOException without propagating.

Can we throw the exception upstream after recording the exception metric ?

---

Updated